### PR TITLE
Basic BDK functionality

### PR DESCRIPTION
**Pull request by benthecarman**
Adds basic bdk functionality like creating the wallet, generating addresses, and syncing.

### CODE REVIEW

File: node-manager/src/wallet.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            Network::Bitcoin => Ok("https://blockstream.info/api"),
+            Network::Testnet => Ok("https://blockstream.info/testnet/api"),
+            Network::Signet => Ok("https://mempool.space/signet/api"),
+            Network::Regtest => Err(bdk::Error::Generic(
+                "No esplora client available for regtest".to_string(),
+            )),
+        }?;
+
+        let blockchain = EsploraBlockchain::new(url, 20);
+        self.wallet.sync(&blockchain, SyncOptions::default()).await
```
Comment (benthecarman) [inline node-manager/src/wallet.rs:49]: this function is only async if the target is `wasm32`, for POC probably fine to ignore but is there something we can do to handle this for different builds?

Comment (futurepaul) [reply node-manager/src/wallet.rs:49]: yeah we'd need an async executor like async-std or tokio if cfg not wasm

Comment (futurepaul) [reply node-manager/src/wallet.rs:49]: if I understand your q

**Comment by TonyGiorgio** [comment]
When I delete my storage and refresh the page, I get a few issues including the browser freezing up.

Here's the console logs:
```
INFO src/nodemanager.rs:55
1. Err(ConnectionClose(CloseEvent { code: 1000, reason: "", was_clean: true })) [node_manager.js:504](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)
DEBUG src/nodemanager.rs:57
WebSocket Closed [node_manager.js:507](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)
Download the React DevTools for a better development experience: https://reactjs.org/link/react-devtools [react-dom.development.js:29840](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/frontend/node_modules/react-dom/cjs/react-dom.development.js)
DEBUG src/lib.rs:27
Main function begins [node_manager.js:507](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)
DEBUG src/lib.rs:28
Main function ends [node_manager.js:507](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)
DEBUG src/lib.rs:27
Main function begins [node_manager.js:507](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)
DEBUG src/lib.rs:28
Main function ends [node_manager.js:507](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)
INFO /home/dev/.cargo/git/checkouts/bdk-914123f374eaebe5/c7a43d9/src/wallet/mod.rs:1380
Derivation of 100 addresses from 0 took 129 ms [node_manager.js:504](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)
DEBUG /home/dev/.cargo/git/checkouts/bdk-914123f374eaebe5/c7a43d9/src/wallet/mod.rs:1682
Begin sync... [node_manager.js:507](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)
DEBUG /home/dev/.cargo/git/checkouts/bdk-914123f374eaebe5/c7a43d9/src/wallet/mod.rs:362
max_address 100 [node_manager.js:507](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)
DEBUG /home/dev/.cargo/git/checkouts/bdk-914123f374eaebe5/c7a43d9/src/wallet/mod.rs:386
caching internal addresses [node_manager.js:507](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)
INFO /home/dev/.cargo/git/checkouts/bdk-914123f374eaebe5/c7a43d9/src/wallet/mod.rs:1380
Derivation of 100 addresses from 0 took 202 ms [node_manager.js:504](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)
DEBUG /home/dev/.cargo/git/checkouts/bdk-914123f374eaebe5/c7a43d9/src/wallet/mod.rs:1694
run_setup: true [node_manager.js:507](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)






XHRGEThttps://blockstream.info/testnet/api/scripthash/3c7b90328b574d62291f1b2db39882ad221ef9b5a935d18a86994b13aa15d822/txs
Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://blockstream.info/testnet/api/scripthash/3c7b90328b574d62291f1b2db39882ad221ef9b5a935d18a86994b13aa15d822/txs. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 502.

panicked at 'Wallet failed to sync: Esplora(Reqwest(reqwest::Error { kind: Request, source: "JsValue(TypeError: NetworkError when attempting to fetch resource.\n)" }))', src/nodemanager.rs:90:34

Stack:

../node-manager/pkg/node_manager.js/getImports/imports.wbg.__wbg_new_abda76e883ba8a5f@http://localhost:3000/static/js/bundle.js:40240:17
console_error_panic_hook::hook::h01517403bb526593@http://localhost:3000/static/media/node_manager_bg.aae36c67166cbac24056.wasm:wasm-function[849]:0xb75c3
core::ops::function::Fn::call::hfe35ce3721f5bdc2@http://localhost:3000/static/media/node_manager_bg.aae36c67166cbac24056.wasm:wasm-function[3758]:0x10231e
std::panicking::rust_panic_with_hook::hda0843f98ee6b6ed@http://localhost:3000/static/media/node_manager_bg.aae36c67166cbac24056.wasm:wasm-function[1296]:0xd2ce7
std::panicking::begin_panic_handler::{{closure}}::h02c1498396c86bcc@http://localhost:3000/static/media/node_manager_bg.aae36c67166cbac24056.wasm:wasm-function[1635]:0xe1e87
std::sys_common:…
[node_manager.js:388](http://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)
Uncaught (in promise) RuntimeError: unreachable executed
    __wbg_adapter_37 node_manager.js:220
    real node_manager.js:193
    promise callback*../node-manager/pkg/node_manager.js/getImports/imports.wbg.__wbg_then_11f7a54d67b4bfad node_manager.js:739
    __wbg_adapter_37 node_manager.js:220
    real node_manager.js:193
    promise callback*../node-manager/pkg/node_manager.js/getImports/imports.wbg.__wbg_then_cedad20fbbd9418a node_manager.js:743
    __wbg_adapter_37 node_manager.js:220
    real node_manager.js:193
    promise callback*../node-manager/pkg/node_manager.js/getImports/imports.wbg.__wbg_then_11f7a54d67b4bfad node_manager.js:739
```

It freezes between `run_setup: true` and the rest. If I refresh the page, I do not have an issue.

**Comment by benthecarman** [comment]
removed the `bdk-macros` stuff, this should be good to go now

**Comment by TonyGiorgio** [comment]
Still no good using ssl

```
panicked at 'already borrowed: BorrowMutError', /home/dev/.cargo/git/checkouts/bdk-914123f374eaebe5/c7a43d9/src/wallet/mod.rs:1320:35

Stack:

../node-manager/pkg/node_manager.js/getImports/imports.wbg.__wbg_new_abda76e883ba8a5f@https://localhost:3000/static/js/bundle.js:40240:17
console_error_panic_hook::hook::h01517403bb526593@https://localhost:3000/static/media/node_manager_bg.aae36c67166cbac24056.wasm:wasm-function[849]:0xb75c3
core::ops::function::Fn::call::hfe35ce3721f5bdc2@https://localhost:3000/static/media/node_manager_bg.aae36c67166cbac24056.wasm:wasm-function[3758]:0x10231e
std::panicking::rust_panic_with_hook::hda0843f98ee6b6ed@https://localhost:3000/static/media/node_manager_bg.aae36c67166cbac24056.wasm:wasm-function[1296]:0xd2ce7
std::panicking::begin_panic_handler::{{closure}}::h02c1498396c86bcc@https://localhost:3000/static/media/node_manager_bg.aae36c67166cbac24056.wasm:wasm-function[1635]:0xe1e87
std::sys_common::backtrace::__rust_end_short_backtrace::he908bb5426995811@https://localhost:3000/static/media/node_manager_bg.aae36c67166cbac24056.wasm:wasm-function[2812]:0xfd012
rust_begin_unwind@https://localhost:3000/static/media/node_manager_bg.aae36c67166cbac24056.wasm:wasm-function[2232]:0xf37cf
core::panicking::panic_fmt::h3684668e1a64c95c@https://localhost:3000/static/media/node_manager_bg.aae36c67166cbac24056.wasm:wasm-function[2619]:0xfa81b
core::result::unwrap_failed::h2a36be58ca648a0c@https://localhost:3000/static/media/node_manager_bg.aae36c67166cbac24056.wasm:wasm-function[1859]:0xe9c2f
nodemanager_get_new_address@https://localhost:3000/static/media/node_manager_bg.aae36c67166cbac24056.wasm:wasm-function[245]:0x67d10
get_new_address@https://localhost:3000/static/js/bundle.js:40173:12
onClick@https://localhost:3000/static/js/bundle.js:198:51
callCallback@https://localhost:3000/static/js/bundle.js:10964:18
invokeGuardedCallbackDev@https://localhost:3000/static/js/bundle.js:11008:20
invokeGuardedCallback@https://localhost:3000/static/js/bundle.js:11065:35
invokeGuardedCallbackAndCatchFirstError@https://localhost:3000/static/js/bundle.js:11079:29
executeDispatch@https://localhost:3000/static/js/bundle.js:15223:46
processDispatchQueueItemsInOrder@https://localhost:3000/static/js/bundle.js:15249:26
processDispatchQueue@https://localhost:3000/static/js/bundle.js:15260:41
dispatchEventsForPlugins@https://localhost:3000/static/js/bundle.js:15269:27
./node_modules/react-dom/cjs/react-dom.development.js/dispatchEventForPluginEventSystem/<@https://localhost:3000/static/js/bundle.js:15429:16
batchedUpdates$1@https://localhost:3000/static/js/bundle.js:29821:16
batchedUpdates@https://localhost:3000/static/js/bundle.js:10812:16
dispatchEventForPluginEventSystem@https://localhost:3000/static/js/bundle.js:15428:21
dispatchEventWithEnableCapturePhaseSelectiveHydrationWithoutDiscreteEventReplay@https://localhost:3000/static/js/bundle.js:12934:42
dispatchEvent@https://localhost:3000/static/js/bundle.js:12928:88
dispatchDiscreteEvent@https://localhost:3000/static/js/bundle.js:12905:22


[node_manager.js:388](https://localhost:3000/data/Dev/BitcoinDevShop/mutiny-web-poc/node-manager/pkg/node_manager.js)
Uncaught RuntimeError: unreachable executed
    get_new_address node_manager.js:304
    onClick App.tsx:60
    React **15**
```

**Comment by TonyGiorgio** [comment]
Also if you try to generate an address while it is syncing, I get this:

```
Uncaught RuntimeError: unreachable executed
    get_new_address node_manager.js:304
    onClick App.tsx:60
    React 23
    tsx index.tsx:7
    factory react refresh:6
    Webpack 3
```

File: node-manager/src/wallet.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        let wallet = Wallet::new(
+            receive_descriptor_template,
+            Some(change_descriptor_template),
+            network,
+            db,
+        )
+        .expect("Error creating wallet");
+
+        MutinyWallet {
+            wallet: Mutex::new(wallet),
```
Comment (TonyGiorgio) [inline node-manager/src/wallet.rs:39]: Does this have to be in a mutex? Every time syncing happens, which can take many seconds, every wallet actions will be blocked. Creating an address for instance isn't possible while syncing is happening. We can fix later but wanted to throw that question out there.

Comment (benthecarman) [reply node-manager/src/wallet.rs:39]: A lot of the actions call `borrow_mut()` on the database which will panic if the database is being used elsewhere (ie syncing). We can change to a non-async mutex and then give an generic error message or remove it and let bdk panic instead

**Review by TonyGiorgio — APPROVED** [review]

**Pull request merged**